src/music_gui.cpp

```diff
@@ -35,278 +36,388 @@
+void MusicSystem::CheckStatus()
+{
+	if ((_game_mode == GM_MENU) != (this->selected_playlist == PLCH_THEMEONLY)) {
+		/* Make sure the theme-only playlist is active when on the title screen, and not during gameplay */
+		this->ChangePlaylist((_game_mode == GM_MENU) ? PLCH_THEMEONLY : (PlaylistChoices)_settings_client.music.playlist);
 	}
-	songinfo.filename = filename; // non-owned pointer
-	songinfo.loop = (_game_mode == GM_MENU) && (_music_wnd_cursong == 1);
-	MusicDriver::GetInstance()->PlaySong(songinfo);
-	SetWindowDirty(WC_MUSIC_WINDOW, 0);
+	if (this->active_playlist.empty()) return;
+	if (this->IsPlaying() && !MusicDriver::GetInstance()->IsSongPlaying()) this->Next();
 }
 
-static void DoStopMusic()
+/** Is the player getting music right now? */
+bool MusicSystem::IsPlaying() const
 {
-	MusicDriver::GetInstance()->StopSong();
-	SetWindowDirty(WC_MUSIC_WINDOW, 0);
+	return _settings_client.music.playing && this->active_playlist.size() > 0;
 }
 
-/** Reload the active playlist data from playlist selection and shuffle setting */
-static void ResetPlaylist()
+/** Is shuffle mode enabled? */
+bool MusicSystem::IsShuffle() const
 {
-	uint i = 0;
-	uint j = 0;
-
-	memset(_cur_playlist, 0, sizeof(_cur_playlist));
-	do {
-		/* File is the index into the file table of the music set. The play list uses 0 as 'no entry',
-		 * so we need to subtract 1. In case of 'no entry' (file = -1), just skip adding it outright. */
-		int file = _playlists[_settings_client.music.playlist][i] - 1;
-		if (file >= 0) {
-			const char *filename = BaseMusic::GetUsedSet()->files[file].filename;
-			/* We are now checking for the existence of that file prior
-			 * to add it to the list of available songs */
-			if (!StrEmpty(filename) && FioCheckFileExists(filename, BASESET_DIR)) {
-				_cur_playlist[j] = _playlists[_settings_client.music.playlist][i];
-				j++;
-			}
-		}
-	} while (_playlists[_settings_client.music.playlist][++i] != 0 && j < lengthof(_cur_playlist) - 1);
-
-	/* Do not shuffle when on the intro-start window, as the song to play has to be the original TTD Theme*/
-	if (_settings_client.music.shuffle && _game_mode != GM_MENU) {
-		i = 500;
-		do {
-			uint32 r = InteractiveRandom();
-			byte *a = &_cur_playlist[GB(r, 0, 5)];
-			byte *b = &_cur_playlist[GB(r, 8, 5)];
-
-			if (*a != 0 && *b != 0) {
-				byte t = *a;
-				*a = *b;
-				*b = t;
-			}
-		} while (--i);
-	}
+	return _settings_client.music.shuffle;
 }
 
-static void StopMusic()
+/** Return the current song, or a dummy if none */
+MusicSystem::PlaylistEntry MusicSystem::GetCurrentSong() const
+{
+	if (!this->IsPlaying()) return PlaylistEntry(BaseMusic::GetUsedSet(), 0);
+	return this->active_playlist[this->playlist_position];
+}
+
+/** Is one of the custom playlists selected? */
+bool MusicSystem::IsCustomPlaylist() const
+{
+	if (this->selected_playlist == PLCH_CUSTOM1) return true;
+	if (this->selected_playlist == PLCH_CUSTOM2) return true;
+	return false;
```

Comment:
`return this->selected_playlist == PLCH_CUSTOM1 || this->selected_playlist == PLCH_CUSTOM2;` ?